### ISSUE TITLE
Deb-Light: Python Path Calculation for MythTV Bindings

### ISSUE THREAD

**Issue by rcrdnalor** [MythTV/packaging#159]
For unprivileged users, Debian changed the way of defining the
location when installing a python module:
In python 3.10, it changed the "posix_prefix" scheme which defaults now to   
"$HOME/lib/python3.10/site-packages" instead of "/usr/lib/python3.8/dist-packages"
This is used in 
https://github.com/MythTV/packaging/blob/79f50b9e532277da16076401c75f2de01b89baf7/deb-light/test.sh#L36
and in
https://github.com/MythTV/packaging/blob/79f50b9e532277da16076401c75f2de01b89baf7/deb-light/config.sh#L93
 
Ubuntu 20.04, based on python3.8 shows:
```
$ python3 -m sysconfig
Platform: "linux-x86_64"
Python version: "3.8"
Current installation scheme: "posix_prefix"

Paths:
        data = "/usr"
        include = "/usr/include/python3.8"
        platinclude = "/usr/include/python3.8"
        platlib = "/usr/lib/python3.8/site-packages"
        platstdlib = "/usr/lib/python3.8"
        purelib = "/usr/lib/python3.8/site-packages"
        scripts = "/usr/bin"
        stdlib = "/usr/lib/python3.8"
```

Ubuntu 22.04, based on python 3.10 shows:
```
local-admin@coretest:~$ python3 -m sysconfig
Platform: "linux-x86_64"
Python version: "3.10"
Current installation scheme: "posix_prefix"

Paths: 
	data = "/home/local-admin"
	include = "/usr/include/python3.10"
	platinclude = "/usr/include/python3.10"
	platlib = "/home/local-admin/lib/python3.10/site-packages"
	platstdlib = "/home/local-admin/lib/python3.10"
	purelib = "/home/local-admin/lib/python3.10/site-packages"
	scripts = "/home/local-admin/bin"
	stdlib = "/usr/lib/python3.10"        
```

MythTVs Python bindings are of type "purelib", thus the following
code can be used to determine the correct path:
```
#!/bin/bash

python_bindings_path=`python3 -m sysconfig | grep "purelib" | cut -d \" -f2`
echo $python_bindings_path
```
which results on Ubuntu 22.04 with python 3.10 (as user 'local-admin') to
```
/home/local-admin/lib/python3.10/site-packages
```

As a consequence, calling python scripts or fetching metadata does not work when using 
`test.sh` running MythTVs frontend or backend.

**Comment by bennettpeter** [comment]
Please help with some information.

I am still on ubuntu 20.04. When I install mythtv in my home directory using packaging/deb-light/install.sh, the python packages end up in /home/peter/proj/build/jammy-amd64/mythtv/master-tst/usr/local/lib/python3.10/dist-packages

When I run python3 -m sysconfig | grep "purelib" | cut -d \" -f2 on 20.04 I get this
/usr/local/lib/python3.10/dist-packages

Where do the python packages end up when you install in your home directory on ubuntu 22.04 using deb-light/install.sh?

If they end up in /home/local-admin/lib/python3.10/site-packages, that will make it difficult to have different versions in your home directory at the same time. It needs to be under the /home/peter/proj/build/jammy-amd64/mythtv/master-tst directory.

What is the value for purelib when running under root?
What is the value for purelib when running under fakeroot?

Perhaps I will need to use fakeroot in the install.sh to get stuff where I want it.

**Comment by rcrdnalor** [comment]
Ubuntu 22.04:

> Where do the python packages end up when you install in your home directory on ubuntu 22.04 using deb-light/install.sh?

Installation of python bindings under deb-light (via install.sh): 
home/local-admin/MythTV/mythtv-light/proj/build/jammy-amd64/mythtv/master/home/local-admin/lib/python3.10/site-packages/MythTV

The corresponding path for 'purelib' is
"/home/local-admin/lib/python3.10/site-packages"

My build name:
$ cat .buildnames 
shortname=ja4mtvmr longname=jammy-amd64/mythtv/master

The build destination:
$ cat .buildrc
BUILDDEST=/home/local-admin/MythTV/mythtv-light/proj/build

> What is the value for purelib when running under root?

Running as root user, I get the path for 'purelib':
"/usr/local/lib/python3.10/dist-packages"

> What is the value for purelib when running under fakeroot?

Running with fakeroot does not produce the expected reault:
"/home/local-admin/lib/python3.10/site-packages"
Same path as without using fakeroot.

In other words, the installation with 'install.sh' produces a unique directory for the MythTVs python bindings,
but the expected path is different on Ubuntu 20.04 and 22.04.

BTW, Ubuntu 20.04 has python3.8 as default.
Depending *how* you installed python3.10, you may not get the 'officially' patched version of python - sysconfig.

**Comment by bennettpeter** [comment]
Actually, I made a mistake there. I actually am on 22.04, jammy. However i still have the old paths:

$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.2 LTS
Release:    22.04
Codename:   jammy

$ python3 -m sysconfig
Platform: "linux-x86_64"
Python version: "3.10"
Current installation scheme: "posix_local"

Paths:
        data = "/usr/local"
        include = "/usr/include/python3.10"
        platinclude = "/usr/include/python3.10"
        platlib = "/usr/local/lib/python3.10/dist-packages"
        platstdlib = "/usr/lib/python3.10"
        purelib = "/usr/local/lib/python3.10/dist-packages"
        scripts = "/usr/local/bin"
        stdlib = "/usr/lib/python3.10"

How do I get a system with the new paths? Maybe 23.04?

Existing PYTHONPATH line in my scripts:
```
export PYTHONPATH=`ls -d $basedir/local/lib/python*/dist-packages`
```

I think all I need for the test.sh script is replace that line with this
```
python_bindings_path=`python3 -m sysconfig | grep "purelib" | cut -d \" -f2`
export PYTHONPATH=$destdir/$python_bindings_path
```
It gives the same result as before on my system with the dist-packages.

However, my deb-light deb package will be installing in the wrong place. It will need to find the correct path at install time.

**Comment by rcrdnalor** [comment]
Hi Peter,
it turned out it is a 'PEBKAC' on my side.
For whatever reasons I had installed python3 locally:
```
$ which python3
/home/local-admin/bin/python3
```
This explains the difference in the calculation of the path for `purelib`.
I will revert that installation.
The pristine builder user for Ubuntu Jammy and Lunar returns:
```
chroot-builder@coretest:~$ schroot -c crblunar
(crblunar)chroot-builder@coretest:~$ 
(crblunar)chroot-builder@coretest:~$ 
(crblunar)chroot-builder@coretest:~$ python3 --version
Python 3.11.2

(crblunar)chroot-builder@coretest:~$ which python3
/usr/bin/python3
(crblunar)chroot-builder@coretest:~$ python3 -m sysconfig
Platform: "linux-x86_64"
Python version: "3.11"
Current installation scheme: "posix_local"

Paths: 
	data = "/usr/local"
	include = "/usr/include/python3.11"
	platinclude = "/usr/include/python3.11"
	platlib = "/usr/local/lib/python3.11/dist-packages"
	platstdlib = "/usr/lib/python3.11"
	purelib = "/usr/local/lib/python3.11/dist-packages"
	scripts = "/usr/local/bin"
	stdlib = "/usr/lib/python3.11"

chroot-builder@coretest:~$ exit
################################################################


chroot-builder@coretest:~$ which python3
/usr/bin/python3

chroot-builder@coretest:~$ python3 --version
Python 3.10.6

chroot-builder@coretest:~$ python3 -m sysconfig
Platform: "linux-x86_64"
Python version: "3.10"
Current installation scheme: "posix_local"

Paths: 
	data = "/usr/local"
	include = "/usr/include/python3.10"
	platinclude = "/usr/include/python3.10"
	platlib = "/usr/local/lib/python3.10/dist-packages"
	platstdlib = "/usr/lib/python3.10"
	purelib = "/usr/local/lib/python3.10/dist-packages"
	scripts = "/usr/local/bin"
	stdlib = "/usr/lib/python3.10"
```
which is exactly the same what you are getting.
Please excuse the noise and a big "Thank You" for pointing me to the right direction.

Roland
